src/dist/replication/meta_server/meta_bulk_load_service.cpp

```diff
@@ -29,6 +30,232 @@ void bulk_load_service::initialize_bulk_load_service()
+            }
+            {
+                zauto_write_lock l(_lock);
+                _bulk_load_app_id.insert(app->app_id);
+                _apps_in_progress_count[app->app_id] = app->partition_count;
+            }
+            create_app_bulk_load_dir(
+                app->app_name, app->app_id, app->partition_count, std::move(rpc));
+        });
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void bulk_load_service::create_app_bulk_load_dir(const std::string &app_name,
+                                                 int32_t app_id,
+                                                 int32_t partition_count,
+                                                 start_bulk_load_rpc rpc)
+{
+    const auto req = rpc.request();
+
+    app_bulk_load_info ainfo;
+    ainfo.app_id = app_id;
+    ainfo.app_name = app_name;
+    ainfo.partition_count = partition_count;
+    ainfo.status = bulk_load_status::BLS_DOWNLOADING;
+    ainfo.cluster_name = req.cluster_name;
+    ainfo.file_provider_type = req.file_provider_type;
+    blob value = dsn::json::json_forwarder<app_bulk_load_info>::encode(ainfo);
+
+    _meta_svc->get_meta_storage()->create_node(
+        get_app_bulk_load_path(app_id), std::move(value), [rpc, ainfo, this]() {
+            dinfo_f("create app({}) bulk load dir", ainfo.app_name);
+            {
+                zauto_write_lock l(_lock);
+                _app_bulk_load_info[ainfo.app_id] = ainfo;
+                _apps_pending_sync_flag[ainfo.app_id] = false;
+            }
+            for (int32_t i = 0; i < ainfo.partition_count; ++i) {
+                create_partition_bulk_load_dir(
+                    ainfo.app_name, gpid(ainfo.app_id, i), ainfo.partition_count, std::move(rpc));
+            }
+        });
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void bulk_load_service::create_partition_bulk_load_dir(const std::string &app_name,
+                                                       const gpid &pid,
+                                                       int32_t partition_count,
+                                                       start_bulk_load_rpc rpc)
+{
+    partition_bulk_load_info pinfo;
+    pinfo.status = bulk_load_status::BLS_DOWNLOADING;
+    blob value = dsn::json::json_forwarder<partition_bulk_load_info>::encode(pinfo);
+
+    _meta_svc->get_meta_storage()->create_node(
+        get_partition_bulk_load_path(pid),
+        std::move(value),
+        [app_name, pid, partition_count, rpc, pinfo, this]() {
+            dinfo_f("app({}) create partition({}) bulk_load_info", app_name, pid.to_string());
+            {
+                zauto_write_lock l(_lock);
+                _partition_bulk_load_info[pid] = pinfo;
+                _partitions_pending_sync_flag[pid] = false;
+                if (--_apps_in_progress_count[pid.get_app_id()] == 0) {
+                    ddebug_f("app({}) start bulk load succeed", app_name);
+                    _apps_in_progress_count[pid.get_app_id()] = partition_count;
```

Comment:
Why `_apps_in_progress_count[pid.get_app_id()]` is set to `partition_count` here?

Reply:
In bulk load process, app bulk load status will be updated when all partitions' bulk load status updated. For example, app bulk load status will transfer from downloading to downloaded only when all partitions bulk load status have transferred to downloaded. `_apps_in_progress_count[app_id]` implies how many partitions have not transferred status to next stage.